### PR TITLE
box: support tuple formats in IPROTO request responses

### LINKED ISSUE #8147: Provide names in IPROTO responses
Now if a user wants to know selected tuples' format, they have to select spaces format manually and maintain its relevance using SCHEMA_VERSION mechanism in IPROTO. Note that this approach is durable only for direct IPROTO DML, such as replaces or selects, while in general doesn't work for IPROTO calls.

Let's provide an optional field(s) in response that describes format of tuple in the response. We should discuss the details - what exactly must be placed in that field(s) and how optional it should be.

### LINKED ISSUE #8633: Support tuple formats in IPROTO call and eval request arguments
For symmetry with #8147, we want to allow users to send tuple arguments in `IPROTO_TUPLE` field of call and eval requests encoded as `MP_TUPLE` with their formats encoded in the same way as formats of tuple results in #8147.

### PR DESCRIPTION
**Pull request by CuriousGeorgiy**
Closes #8147
Closes #8633

### CODE REVIEW

**Comment by CuriousGeorgiy** [comment]
Please note that tests using `call_16` are failing right now, because this patch doesn't work with it — I will discuss `call_16` in the context of this patch with @locker and @alyapunov on Wednesday.

**Comment by CuriousGeorgiy** [comment]
@locker regarding checkpatch linter errors:

1. Ignoring errors for #8631 patch.
2. Ignoring errors for uncommented defitions of functions, since they didn't have comments prior to this patch.

**Review by alyapunov — CHANGES_REQUESTED** [review]

File: src/box/lua/init.c [hunk 1]
```diff
@@ -682,21 +682,25 @@ static const struct luaL_Reg boxlib_backup[] = {
 /**
  * A MsgPack extensions handler, for types defined in box.
  */
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/lua/init.c:688]: Fixed. Also added a comment to `luamp_encode_extension_f`.

File: src/box/mp_tuple.c [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	mp_decode_extl(data, &type);
+	if (type != MP_TUPLE || tuple_unpack(data, format_map, tuple) == NULL) {
+		*data = svp;
+		return NULL;
+	}
+	return *tuple;
+}
+
+struct tuple *
+tuple_unpack(const char **data, struct tuple_format_map *format_map,
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/mp_tuple.c]: Fixed. Also fixed `mp_decode_tuple`.

File: src/box/mp_tuple.c [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	return mp_sizeof_ext(tuple_len(tuple));
+}
+
+/** Pack a tuple value to a buffer. */
+static char *
+tuple_pack(char *data, struct tuple *tuple)
+{
+	data = mp_encode_uint(data, tuple->format_id);
+	uint32_t bsize = tuple_bsize(tuple);
+	tuple_to_buf(tuple, data, bsize);
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/mp_tuple.c]: Fixed.

File: src/box/mp_tuple.h [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+/** Return the number of bytes an encoded tuple value takes. */
+uint32_t
+mp_sizeof_tuple(struct tuple *tuple);
+
+/**
+ * Decode a tuple value from MsgPack data.
+ *
+ * @param data A buffer.
+ * @param idx Tuple format index.
+ * @param ctx MsgPack decoding context.
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/mp_tuple.h]: Fixed. Also fixed `tuple_unpack`.

File: src/lib/core/mp_ctx.h [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+	 */
+	struct mh_strnu32_t *translation;
+	/** 'Virtual' destructor. */
+	void (*destroy)(struct mp_ctx *ctx);
+};
+
+static inline void
+mp_ctx_destroy(struct mp_ctx *ctx)
+{
+	ctx->destroy(ctx);
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/lib/core/mp_ctx.h]: Fixed.

File: src/lib/core/mp_ctx.h [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+ * Base context for MsgPack encoding/decoding.
+ */
+struct mp_ctx {
+	/**
+	 * If a  first-level `MP_MAP` key has `MP_STRING` type, the key is
+	 * looked up and replaced with a translation, if found. The translation
+	 * table must use `lua_hash` as the hash function.
+	 */
+	struct mh_strnu32_t *translation;
+	/** 'Virtual' destructor. */
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/lib/core/mp_ctx.h]: Fixed.

File: src/box/tuple_format_map.c [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+struct tuple_format *
+tuple_format_map_find(struct tuple_format_map *map, uint16_t format_id)
+{
+	if (map->cache_last_index == -1) {
+		assert(map->hash_table == NULL);
+		return NULL;
+	}
+	for (size_t i = 0; i < TUPLE_FORMAT_MAP_CACHE_SIZE; ++i) {
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/tuple_format_map.c]: Fixed, thanks.

File: src/box/tuple_format_map.c [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+		map->cache[0] = (struct mh_i32ptr_node_t){
+			.key = format_id,
+			.val = format,
+		};
+		map->cache_last_index = 0;
+		return;
+	}
+	if (map->hash_table == NULL) {
+		if (map->cache_last_index < TUPLE_FORMAT_MAP_CACHE_SIZE - 1) {
+			map->cache[++map->cache_last_index] =
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/tuple_format_map.c]: Fixed.

File: src/box/tuple_format_map.c [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+/**
+ * Add a tuple format to the tuple format and reference the format.
+ */
+static void
+tuple_format_map_add_format_impl(struct tuple_format_map *map,
+				 uint16_t format_id,
+				 struct tuple_format *format)
+{
+	tuple_format_ref(format);
+	if (map->cache_last_index == -1) {
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/tuple_format_map.c]: Fixed.

File: src/box/lua/iproto.c [hunk 10]
```diff
@@ -264,17 +269,20 @@ lbox_iproto_send(struct lua_State *L)
...
 	if (L == NULL)
 		return IPROTO_HANDLER_ERROR;
-	int cb_ref = (int)(uintptr_t)ctx;
+	int cb_ref = (int)(uintptr_t)cb_ctx;
 	lua_rawgeti(L, LUA_REGISTRYINDEX, cb_ref);
-	luamp_push_with_translation(L, header, header_end,
-				    iproto_key_translation);
-	luamp_push_with_translation(L, body, body_end,
-				    iproto_key_translation);
+	struct mp_ctx *mp_ctx = xmalloc(sizeof(*mp_ctx));
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/lua/iproto.c]: Fixed.

[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/lua/iproto.c]: Fixed.

**Review by alyapunov — CHANGES_REQUESTED** [review]

File: src/box/iproto.cc [hunk 11]
```diff
@@ -2199,10 +2205,24 @@ tx_process1(struct cmsg *m)
...
+	    tuple_format_map_to_iproto_obuf(&format_map, out) != 0) {
 		goto error;
+	}
+	tuple_format_map_destroy(&format_map);
 	iproto_reply_select(out, &svp, msg->header.sync, ::schema_version,
-			    tuple != 0);
+			    tuple != 0, box_tuple_as_ext);
 	iproto_wpos_create(&msg->wpos, out);
 	tx_end_msg(msg, &svp);
 	return;
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/iproto.cc:2185]: Fixed.

File: src/box/iproto.cc [hunk 12]
```diff
@@ -2374,21 +2422,46 @@ tx_process_call(struct cmsg *m)
...
+		goto discard;
+	} else {
+		int rc = 0;
+		if (box_tuple_as_ext) {
+			struct tuple_format_map *tuple_format_map =
+				&ctx.tuple_format_map;
+			rc = tuple_format_map_to_iproto_obuf(tuple_format_map,
+							     out);
+		}
+		if (box_tuple_as_ext)
```
[the remark this answers is not in the document]
Comment (CuriousGeorgiy) [reply src/box/iproto.cc:2399]: Fixed.

**Comment by CuriousGeorgiy** [comment]
@locker lint/checkpatch seems to be FP on this hunk:

```diff
diff --git a/src/lib/core/mp_ctx.h b/src/lib/core/mp_ctx.h
index 6230b2795..33d6aea89 100644
--- a/src/lib/core/mp_ctx.h
+++ b/src/lib/core/mp_ctx.h
@@ -36,17 +44,52 @@ struct mp_ctx {
 };

 static inline void
-mp_ctx_create(struct mp_ctx *ctx, struct mh_strnu32_t *translation)
+mp_ctx_create(struct mp_ctx *ctx, struct mh_strnu32_t *translation,
+             void (*destroy)(struct mp_ctx *),
+             void (*move)(struct mp_ctx *, struct mp_ctx *))
 {
        ctx->translation = translation;
+       ctx->destroy = destroy;
+       ctx->move = move;
+}
```

**Review by ochaplashkin — APPROVED** [review]
LGTM :rocket:

**Review by alyapunov — APPROVED** [review]

**Comment by Totktonada** [comment]
OK regarding 'lua: handle MP_TUPLE extension in merge buffer source's next method' commit.

**Pull request merged**